src/services/DomainModel.ts

```diff
@@ -0,0 +1,125 @@
+// tree that we don't care about in the context of an active Query.
+
+export const unknown = Symbol("Unknown");
+export type Unknown = typeof unknown;
+
+// These `...ID` types prevent `let gatewayID = getNodeID()` which would sadly
+// compile if each were simply string. Is there an easier way? Is it worth it?
+export type GatewayID = { type: "GatewayID"; id: string };
+export type NodeID = { type: "NodeID"; id: string };
+export type ReadingSchemaID = { type: "ReadingSchemaID"; id: string };
+
+// TODO (carl): So the types don't need to be peppered with `| Unknown`:
+//
+// - we might be able to make a type like Shallow<Node> that's not guaranteed to
+//   link to every other object. I doubt it would be worth the complication.
+//
+// - we could just use `?` and `| undefined` and explain that undefined means
+//   something like "idk" or "Unknown". I just hate comments explaining
+//   things when good naming will make semantics obvious. Is it a common enough
+//   practice in JS? For example `sensors?: Set<Sensor | undefined>;`
+
+export interface Gateway {
+  // Attributes
+  name: string;
+  description: string;
+  shortDescription: string;
+  // Links
+  nodes: Set<Node> | Unknown;
+  // Implementation Details
+  readonly id: GatewayID; // Reference across the whole system architecture
+  type: "Gateway"; // For typescript narrowing and javascript debugging
+}
+
+export interface Node {
+  // Attributes
+  description: string;
+  shortDescription: string;
+  locationHref: string;
+  locationName: string;
+  name: string;
+  // Links
+  gateway: Gateway;
+  sensors: Set<Sensor> | Unknown;
+  // Implementation Details
+  readonly id: NodeID;
+  type: "Node";
+}
+
+export interface Sensor {
+  // Attributes
+  // Links
+  node: Node;
+  readings: Set<Reading> | Unknown;
+  schema: ReadingSchema;
+  // Implementation Details
+  type: "Sensor";
+}
+
+export interface ReadingSchema {
+  // Attributes
+  measure: string; // Temperature, etc.
+  name: string; // Outdoor Temperature, etc.
+  systemOfMeasurement: string; // SI, Imperial, US Customary, etc.
+  unit: string; // Kelvin, etc.
+  unitSymbol: string; // K, etc.
+  // Links
+  // Implementation Details
+  readonly id: ReadingSchemaID;
+  type: "ReadingSchema";
+}
+
+export interface Reading {
+  // Attributes
+  value: JSONValue;
+  timestamp: Date;
+  // Links
+  schema: ReadingSchema;
+  // Implementation Details
+  type: "Reading";
+}
+
+////////////////////////////////////////////////////////////////////////////////
+// Query Stuff
+//
+// A user of the website might not 'think' about the following things unless
+// they had a way to save favorite queries as first class citizens of the app.
+//
+// TODO (carl): Should we put these in a separate file because of the above?
+
+export type DateRange = { from: Date; to: Date };
+
+export const doNotCare = Symbol("DoNotCare");
+export type DoNotCare = typeof doNotCare;
+
+export type QueryFilter = {
```

Comment:
This is fine for a more generic layer for more flexible queries, but for now we can still use an interface tailored specifically to the needs of the client (webapp front end for now) which may later build on something more flexible (or need the flexibility if we have advanced filtering in the backend.) For now, have to build a platform inside the app for all possible queries...at least not until we need it. (Imagine JSON mock data...that would be hard to build such a general query engine.)

Reply:
done this and more